### ISSUE TITLE
6 month old Docker file with nailed versions no longer runs

### ISSUE THREAD

**Issue by johngrabner** [Lightning-AI/lightning#16840]
### Bug description

I suspect lightning is pulling in libraries with floating (unspecified) revisions, thus making 
a specific revision of lightning insufficient to reproduce a build environment.

On one machine (webserver), this docker image has been running since 2022_06_09.
On another machine (development), I tried to re-run the training using the same docker yaml, now lightning will not start.

The docker yaml only has nailed versions:

```
FROM pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime
RUN pip install pytorch-lightning==1.5.10
```

The error:
```
Traceback (most recent call last):
  File "train.py", line 3, in <module>
    import pytorch_lightning as pl
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/__init__.py", line 20, in <module>
    from pytorch_lightning.callbacks import Callback  # noqa: E402
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/callbacks/__init__.py", line 14, in <module>
    from pytorch_lightning.callbacks.base import Callback
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/callbacks/base.py", line 26, in <module>
    from pytorch_lightning.utilities.types import STEP_OUTPUT
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/types.py", line 25, in <module>
    from torchmetrics import Metric
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/__init__.py", line 14, in <module>
    from torchmetrics import functional  # noqa: E402
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/functional/__init__.py", line 14, in <module>
    from torchmetrics.functional.audio.pit import permutation_invariant_training, pit_permutate
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/functional/audio/__init__.py", line 15, in <module>
    from torchmetrics.functional.audio.sdr import (  # noqa: F401
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/functional/audio/sdr.py", line 28, in <module>
    solve = torch.linalg.solve
AttributeError: module 'torch.linalg' has no attribute 'solve'
```

My concern is twofold:
- how can I specify a reproducible build context if providing a 3 digit version is insufficient?
- how to fix this problem (I assume upgrading either lightning or torch)

### How to reproduce the bug

```python
import pytorch_lightning as pl
```

### Error messages and logs

```
Traceback (most recent call last):
  File "train.py", line 3, in <module>
    import pytorch_lightning as pl
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/__init__.py", line 20, in <module>
    from pytorch_lightning.callbacks import Callback  # noqa: E402
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/callbacks/__init__.py", line 14, in <module>
    from pytorch_lightning.callbacks.base import Callback
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/callbacks/base.py", line 26, in <module>
    from pytorch_lightning.utilities.types import STEP_OUTPUT
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/types.py", line 25, in <module>
    from torchmetrics import Metric
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/__init__.py", line 14, in <module>
    from torchmetrics import functional  # noqa: E402
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/functional/__init__.py", line 14, in <module>
    from torchmetrics.functional.audio.pit import permutation_invariant_training, pit_permutate
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/functional/audio/__init__.py", line 15, in <module>
    from torchmetrics.functional.audio.sdr import (  # noqa: F401
  File "/opt/conda/lib/python3.8/site-packages/torchmetrics/functional/audio/sdr.py", line 28, in <module>
    solve = torch.linalg.solve
AttributeError: module 'torch.linalg' has no attribute 'solve'

```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

**Comment by awaelchli** [comment]
Hey @johngrabner 
Which version of torchmetrics got installed there? It is trying to call `torch.linalg` but that module was only introduced after torch 1.8.0. So I suspect that the torchmetrics version is "too new".

**Comment by awaelchli** [comment]
You could try to downgrade torchmetrics to 0.4.1 if it's too new.

**Comment by Borda** [comment]
The closes configuration with our dockers is
`docker pull pytorchlightning/pytorch_lightning:1.5.3-py3.7-torch1.7`
could you pls check it out?

**Comment by johngrabner** [comment]
I will investigate where the torchmetrics comes from (ie: pulled in by pytorch or pulled in from lightning).
I don't believe pytorchlightning/pytorch_lightning:1.5.3-py3.7-torch1.7 Includes cuda.  The pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime does include cudo, so only thing my host needs is nvidia drivers.

**Comment by johngrabner** [comment]
I figured it out, this is my issue.

To have a reproducible docker-compose environment, I need to explicitly install all the requirements ligthning list in requirements.txt; otherwise, lightning will install some unspecified version that meets it's requirements. 

In this particular case, lightning needs torchmetrics>=0.4.1, and it installs a version that is incompatible with my torch, so my torch got upgraded, the upgraded version caused its own issues.

Bottom line, if I want to have a predictable build, I need to install all the specified requirements before installing lightning.  Sorry for the trouble ticket.

**Comment by awaelchli** [comment]
Thanks for the update @johngrabner. The pip installation upgrade battle, we all know it!
